python/socketioclient/socketio_client.py

```diff
@@ -0,0 +1,142 @@
+            self.subscription_requests.append({
+                'token_pair': {
+                    'base_token': base_token,
+                    'quote_token': quote_token
+                },
+                'quantity': levels,
+                'quantity_token': quote_token,  # only for v2/subscribe (optional)
+                'client_request_id': str(uuid.uuid4()),
+                'echo_id': True
+            })
+
+    def on_connect(self):
+        logger.info('Server connected.')
+        for subscription_request in self.subscription_requests:
+            self.emit('subscribe', subscription_request, namespace='/streaming')
+            client_request_id = subscription_request['client_request_id']
+            self.previous_response_times[client_request_id] = time.time()
+        logger.info('Finished subscribing.')
+
+    # def on_connect(self):
+    #     logger.info('Server connected.')
+    #     config_request = {
+    #         'message_type': "GET_ALLOWED_MARKETS",
+    #         'client_request_id': "5c5325e3-ee42-76fa-932c-64dce446d8be"
+    #     }
+    #     logger.info("Client sent request: " + str(config_request))
+    #     self.emit('request', config_request, namespace='/streaming')
+    #     client_request_id = config_request['client_request_id']
+    #     self.previous_response_times[client_request_id] = time.time()        
+    #     logger.info('Finished call.')
+
+    def on_disconnect(self, *args):
+        logger.info('Server disconnected.' + str(args))
+
+    def on_connect_error(self, *args):
+        logger.info("Cannot connect to the server. Error", args)
+
+    def on_response(self, *args):
+        response_time = time.time()
+        client_request_id = args[0]['client_request_id']
+        logger.info("Client received response: " + str(args))
+        previous_response_time = self.previous_response_times[client_request_id]
+        logger.info(f'Client_request_id: {client_request_id}, time taken for first response {response_time - previous_response_time}')
+        self.previous_response_times[client_request_id] = response_time
+
+    def on_stream(self, *args):
+        response_time = time.time()
+        client_request_id = args[0][0]['client_request_id']
+        previous_response_time = self.previous_response_times[client_request_id]
+        streaming_response_time = response_time - previous_response_time
+        running_avg, num_of_responses = self.running_avg_times.get(client_request_id, (0, 0))
+        running_avg = (running_avg * num_of_responses + streaming_response_time) / (num_of_responses + 1)
+        logger.info("printing arguments ", *args)
+        if num_of_responses % self.print_every_n_response == 0:
+            logger.info(f'client_request_id: {client_request_id}, streaming heart beat: {streaming_response_time}, running average: {running_avg}')
+
+        self.running_avg_times[client_request_id] = (running_avg, num_of_responses + 1)
```

Comment:
all of these calculations are also not needed, this was done only internally to check the response times

Reply:
Done!